### PR TITLE
Add (transitive) imports to compiler package

### PR DESCRIPTION
**Pull request by juliaogris**
Rework compiler package for imports. Create new function
```
compiler.NewFileDescriptorSet(filenames, importPaths, includeImports)
```
and load `.proto` dependencies from file.

### CODE REVIEW

File: compiler/compiler.go [hunk 1]
```diff
@@ -1,210 +1,136 @@
...
-
-func newMessage(m *parser.Message, proto3 bool, scope []string, types types) *pb.DescriptorProto {
-	b := &messageBuilder{
-		proto3: proto3,
-		scope:  append(scope, m.Name),
-		types:  types,
-	}
-	b.buildMessageDescriptor(m)
-	return b.messageDesc
+type Proto struct {
```
Comment (alecthomas) [inline compiler/compiler.go]: Never used  outside the package? Make it private?

File: compiler/compiler.go [hunk 2]
```diff
@@ -1,210 +1,136 @@
...
-	parser.SFixed64: pb.FieldDescriptorProto_TYPE_SFIXED64,
-	parser.Bool:     pb.FieldDescriptorProto_TYPE_BOOL,
-	parser.String:   pb.FieldDescriptorProto_TYPE_STRING,
-	parser.Bytes:    pb.FieldDescriptorProto_TYPE_BYTES,
-}
-
-func (b *messageBuilder) createField(pf *parser.Field) *pb.FieldDescriptorProto {
-	if pf.Direct == nil || pf.Direct.Type.Map != nil {
-		panic(fmt.Sprintf("%s: non-direct not implemented", pf.Pos))
+func readProto(file string, importPaths []string, done map[string]bool) ([]*Proto, error) {
```
Comment (alecthomas) [inline compiler/compiler.go]: I think these functions could use some comments, it's not super clear to me what the difference between them all is.

File: compiler/compiler.go [hunk 3]
```diff
@@ -1,210 +1,136 @@
...
+	protos = append(protos, p)
+	return protos, nil
 }
 
-func proto3Optional(pf *parser.Field, proto3 bool) *bool {
-	if pf.Optional && proto3 {
-		p3Optional := true
-		return &p3Optional
+func search(file string, importPaths []string) (io.ReadCloser, error) {
+	if file == "-" {
```
Comment (alecthomas) [inline compiler/compiler.go]: Is this a good idea to have this deep in the code? I would expect something like this to be handled in the CLI interface.

File: compiler/compiler.go [hunk 4]
```diff
@@ -1,210 +1,136 @@
...
-	for _, s := range ss[1:] {
-		result += strings.Title(strings.ToLower(s))
+	for _, path := range importPaths {
+		fname := filepath.Join(path, file)
+		if f, err := os.Open(fname); err == nil {
+			return f, nil
+		}
 	}
-	return &result
+	return nil, fmt.Errorf("cannot find %q on import importPaths", file)
```
Comment (alecthomas) [inline compiler/compiler.go]: `import importPaths` -> `import paths`?

File: compiler/compiler_test.go [hunk 5]
```diff
@@ -20,50 +19,58 @@ func requireNoError(t *testing.T, err error) {
...
-		t.Errorf("protos not equal (-want +got):\n%s", diff)
-	}
-}
-
 func requireEqual(t *testing.T, want, got interface{}) {
 	t.Helper()
-	opts := cmp.Options{protocmp.Transform()}
-	if diff := cmp.Diff(want, got, opts); diff != "" {
-		t.Errorf("not equal (-want +got):\n%s", diff)
+	var opts []cmp.Option
```
Comment (alecthomas) [inline compiler/compiler_test.go:24]: Should we just bring in testify?

File: compiler/filedescriptor.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+)
+
+type messageBuilder struct {
+	proto3      bool
+	messageDesc *pb.DescriptorProto
+	types       types
+	scope       []string
+}
+
+func NewFileDescriptor(p *Proto, types types) *pb.FileDescriptorProto {
```
Comment (alecthomas) [inline compiler/filedescriptor.go]: This public function shouldn't accept a private type as a parameter. Either make the type public or the function private.

File: compiler/typeanalysis.go [hunk 7]
```diff
@@ -63,17 +63,22 @@ func (t types) addName(relTypeName string, scope []string) {
...
-func analyseTypes(p *parser.Proto) types {
+func newTypes(protos []*Proto) types {
 	t := types{}
+	for _, p := range protos {
+		analyseTypes(p, t)
+	}
+	return t
+}
+
+func analyseTypes(p *Proto, t types) types {
```
Comment (alecthomas) [inline compiler/typeanalysis.go]: Return value seems unnecessary as it mutates "t" in place?

**Review by alecthomas — APPROVED** [review]

File: compiler/compiler.go (hunk 2 above)
Comment (juliaogris) [reply compiler/compiler.go]: Fair. I've tried:
```
// readProtos creates ASTs for given files and their dependencies in
// order of the original files slice. By contrast, readProto creates an
// AST for a given file and the ASTs of its dependencies, listed before
// the file AST. Multiple dependencies(imports) listed in the same
// proto file are again processed in order by readProtos(). Already
// processed files are not double listed.
//
// For example:
// * f1 imports f2 which imports f3
// * g1 imports g2, g3, f3
// readProtos([]string{f1, g1}, ...)
// results in the following order: f3, f2, f1, g2, g3, f2
```

hope that's ok.

File: compiler/compiler.go (hunk 1 above)
Comment (juliaogris) [reply compiler/compiler.go]: yeah - naming is hard. it was private and clashed with `import "google.golang.org/protobuf/proto"`.
i think i am a little rusty in this domain - what would be a good name?
I'll rename it to `ast` - please let me know if that's completely off.

File: compiler/compiler.go (hunk 3 above)
Comment (juliaogris) [reply compiler/compiler.go]: yeah hmm. probably not a great idea. `protoc -o all.pb -` doesn't even work. removed.

File: compiler/compiler_test.go (hunk 5 above)
Comment (juliaogris) [reply compiler/compiler_test.go:24]: Yes. Let me do this in a follow up PR if that's ok.

File: compiler/filedescriptor.go (hunk 6 above)
Comment (juliaogris) [reply compiler/filedescriptor.go]: I'll make the function private for now.

**Comment by juliaogris** [comment]
hey @alecthomas - this is a bigger PR setting up the structure for a bunch more work to come. I've reworked this PR a fair bit according to your suggestions, not sure if you want to take another look before I merge, I'll leave it open for a little while longer.

**Comment by alecthomas** [comment]
Nope ship it! The comments are just for you to digest, fix if you think it's necessary, then merge :)

**Pull request merged**